### PR TITLE
Update tfe architecture

### PR DESCRIPTION
**Pull request by ml4**
- [ ] inaccuracy
- [x] clarification
- [x] new docs
- [ ] cosmetic bug - fixing broken text or markup
- [ ] enhancement - changing the website's behavior/layout

Merging the 2019 TFE internal architecture gdoc into public docs to help customers with increasingly deep and probing questions as adoption rises.

### CODE REVIEW

**Comment by sudomateo** [comment]
Going to run this locally and read through it.

File: content/source/docs/enterprise/system-overview/architecture.html.md [hunk 1]
```diff
@@ -5,24 +5,202 @@ page_title: "Architecture Summary - System Overview - Terraform Enterprise"
...
+|  | `retraced-api` | RetraceD API | API provider for the audit log component of Replicated and communicates internally with the audit log’s Postgres and NSQ services. |
+|  | `retraced-cron` | RetraceD Cron | Used to provide time-based audit message queue processing. |
+|  | `retraced-nsqd` | RetraceD NSQD | This is the queue for the Replicated audit log and is internal to Replicated. |
+|  | `retraced-processor` | RetraceD Processor | Processing facility for the audit log component of Replicated and communicates internally with the audit log’s Postgres and NSQ services |
+|  | `retraced-postgres` | RetraceD PostgreSQL | This is the database for the Replicated audit log and is internal to Replicated. |
+|  | `statsd-graphite` | StatsD | This image is used for a metrics service that runs when the application is running. |
+|  | `support-bundle` | Support Bundle Generator | This image is run to collect system information when the customer creates a support bundle. Not used in Active/Active - use [the CLI](https://www.terraform.io/docs/enterprise/admin/active-active.html#support-bundle) instead. |
```
Comment (amy-hashi) [inline content/source/docs/enterprise/system-overview/architecture.html.md:186]: support-bundle runs in a/a when called, same as standalone.

File: content/source/docs/enterprise/system-overview/architecture.html.md [hunk 2]
```diff
@@ -5,24 +5,202 @@ page_title: "Architecture Summary - System Overview - Terraform Enterprise"
...
+
+
+1. `ptfe_atlas` receives the callback request from `ptfe_nginx` and streams the response to `ptfe_archivist`.
+2. `ptfe_archivist` saves the Sentinel results to blob storage and returns an Object ID to `ptfe_atlas`.
+
+Finally, `ptfe_atlas` updates the metadata in PostgreSQL after it receives confirmation that the details of the run have been persisted.
+
+## Full Container List
+
+The following table lists every container Terraform Enterprise runs via the Replicated Native Scheduler.  The list is presented by layer as per the above document sections.
```
Comment (amy-hashi) [inline content/source/docs/enterprise/system-overview/architecture.html.md:157]: so, this is all the long-running service containers, but not every container. there are a variety of ephemeral containers that run on startup and for batch jobs that will show up with random names.

File: content/source/docs/enterprise/system-overview/architecture.html.md [hunk 3]
```diff
@@ -5,24 +5,202 @@ page_title: "Architecture Summary - System Overview - Terraform Enterprise"
...
+
+1. `ptfe_atlas` sends the Object ID from `ptfe_archivist` to the `ptfe_registry_api`.
+1. `ptfe_registry_api` pulls the module from `ptfe_archivist`, saves the metadata in PostgreSQL, and passes the details to the `ptfe_registry_worker`.
+1. `ptfe_registry_worker` extracts the individual modules from your repository, updates the Private Registry tables in PostgreSQL, and `POST`s the results to `ptfe_atlas`.
+
+Finally, `ptfe_atlas` receives the callback and sets the module’s state to *ingressed* in PostgreSQL. The update is now live in the UI.
+
+### Webhook Receipt
+
+Communications between PTFE and a connected VCS server happen via HTTP webhooks.  The latest state of the connected Git repository is downloaded using the process below, and then followup Terraform commands are applied by the state machine.
```
Comment (amy-hashi) [inline content/source/docs/enterprise/system-overview/architecture.html.md:109]: PTFE - the P is silent..

File: content/source/docs/enterprise/system-overview/architecture.html.md [hunk 4]
```diff
@@ -5,24 +5,202 @@ page_title: "Architecture Summary - System Overview - Terraform Enterprise"
 
 # Terraform Enterprise Architecture
 
-This page describes aspects of the architecture of Terraform Enterprise.
+Terraform Enterprise (TFE) is powered by the same set of services behind HashiCorp’s public SaaS application Terraform Cloud.  The sections below present the functional layers of the architecture, how they interact and details pertaining to the corresponding docker containers involved.  References to Docker containers in this document are those you will see when running a `docker ps` on a TFE application server while the application is up.
```
Comment (amy-hashi) [inline content/source/docs/enterprise/system-overview/architecture.html.md:8]: We generally avoid abbreviating the product name anywhere on our website where we can avoid it. TFE should be spelled out as Terraform Enterprise throughout out website documentation.

**Review by amy-hashi — COMMENTED** [review]
I've got some specific comments, but also, I think we'd like to hold off on publishing this as there are some re-names in flight as well as some architecture changes coming in the semi-near future.

File: content/source/docs/enterprise/system-overview/architecture.html.md (hunk 4 above)
Comment (aaron-lane) [inline content/source/docs/enterprise/system-overview/architecture.html.md:8]: Capitalization fix:
```suggestion
Terraform Enterprise (TFE) is powered by the same set of services behind HashiCorp’s public SaaS application Terraform Cloud. The sections below present the functional layers of the architecture, how they interact and details pertaining to the corresponding Docker containers involved. References to Docker containers in this document are those you will see when running a `docker ps` on a TFE application server while the application is up.
```

File: content/source/docs/enterprise/system-overview/architecture.html.md [hunk 5]
```diff
@@ -5,24 +5,202 @@ page_title: "Architecture Summary - System Overview - Terraform Enterprise"
...
 
--> **Note:** This page was written for a prior implementation of Terraform Enterprise, and may be outdated. We plan to review our architecture documentation in the near future.
+~> **Advanced:** The content in this document is not required to operate TFE but is provided to help advanced users understand the product in more depth.
 
-## Services
+~> **Note:** References to `ptfe` below is a historical reference. See [here](https://www.terraform.io/docs/enterprise/index.html#note-about-product-names) for more details.
```
Comment (aaron-lane) [inline content/source/docs/enterprise/system-overview/architecture.html.md:12]: Wording suggestion:
```suggestion
~> **Note:** Several services described below include the legacy string `ptfe` as part of their names. See [here](https://www.terraform.io/docs/enterprise/index.html#note-about-product-names) for more details.
```

**Review by aaron-lane — CHANGES_REQUESTED** [review]
A couple of small suggestions follow. I don't want to overdo the review since a couple other team members are already on it.

File: content/source/docs/enterprise/system-overview/architecture.html.md [hunk 6]
```diff
@@ -5,24 +5,202 @@ page_title: "Architecture Summary - System Overview - Terraform Enterprise"
...
+
+The Coordination Layer is responsible for queuing asynchronous jobs and distributing them to background services running in the Application Layer.
+
+![tfe-int-coordination-layer](assets/tfe-int-coordination-layer.png)
+
+~> **Note:** While these components are critical to the overall operation of TFE, operators rarely (if ever) need to look at them directly.
+
+## Storage Layer
+
+All application data is stored using PostgreSQL and blob storage. Unless you have configured TFE to use External Services, local instances of these services will be started for you.
```
[the remark this answers is not in the document]
Comment (ml4) [reply content/source/docs/enterprise/system-overview/architecture.html.md:49]: Content updated

File: content/source/docs/enterprise/system-overview/architecture.html.md [hunk 7]
```diff
@@ -5,24 +5,202 @@ page_title: "Architecture Summary - System Overview - Terraform Enterprise"
...
+
+This is the largest part of the Application Layer and includes multi-purpose Sidekiq workers, Go services that provide key functionality of TFE, and a dynamic fleet of isolated execution environments that perform Terraform runs.
+
+## Coordination Layer
+
+The Coordination Layer is responsible for queuing asynchronous jobs and distributing them to background services running in the Application Layer.
+
+![tfe-int-coordination-layer](assets/tfe-int-coordination-layer.png)
+
+~> **Note:** While these components are critical to the overall operation of TFE, operators rarely (if ever) need to look at them directly.
```
[the remark this answers is not in the document]
Comment (ml4) [reply content/source/docs/enterprise/system-overview/architecture.html.md:45]: Content updated

File: content/source/docs/enterprise/system-overview/architecture.html.md [hunk 8]
```diff
@@ -5,24 +5,202 @@ page_title: "Architecture Summary - System Overview - Terraform Enterprise"
...
+![tfe-int-tf-plan-apply](assets/tfe-int-tf-plan-apply.png)
+
+1. `ptfe_atlas` detects a state change and enqueues a background job for `ptfe_sidekiq` in Redis.
+2. `ptfe_sidekiq` asynchronously pulls the job and passes the details to `ptfe_build_manager`.
+3. `ptfe_build_manager` enqueues a `ptfe_build_worker` job in RabbitMQ.
+4. `ptfe_build_worker` spins up a temporary environment, runs a Terraform plan or apply, and passes the results to `ptfe_archivist`.
+5. `ptfe_archivist` persists the plan or apply results and returns a URL to `ptfe_build_worker`.
+6. `ptfe_build_worker` passes the plan or apply results back to `ptfe_build_manager`.
+7. `ptfe_build_manager` `POST`s the results to an internal webhook.
+8. `ptfe_nginx` receives the request and forwards the callback to `ptfe_atlas`.
```
[the remark this answers is not in the document]
Comment (ml4) [reply content/source/docs/enterprise/system-overview/architecture.html.md:131]: I got this content from the original gdoc and am effectively just reformatting it for publication. I redrew the diagrams for the same reason but they come exactly from that doc - step 8 cited corresponds to a point in the image above it so if amendments are required to the other diagrams correspondingly, could you tell me exactly where to make the changes please?

File: content/source/docs/enterprise/system-overview/architecture.html.md (hunk 2 above)
Comment (ml4) [reply content/source/docs/enterprise/system-overview/architecture.html.md:157]: Is it the view of engineering that these non-randomly-named but ephemeral containers should be included in this internal doc?

File: content/source/docs/enterprise/system-overview/architecture.html.md [hunk 9]
```diff
@@ -5,24 +5,201 @@ page_title: "Architecture Summary - System Overview - Terraform Enterprise"
...
+
+The Coordination Layer is responsible for queuing asynchronous jobs and distributing them to background services running in the Application Layer. Standalone deployments will start a local Redis container for this purpose and Active/Active deployments will use an external Redis service.
+
+![tfe-int-coordination-layer](assets/tfe-int-coordination-layer.png)
+
+~> **Note:** While these components are critical to the overall operation of Terraform Enterprise, operators rarely (if ever) need to look at them directly.
+
+## Storage Layer
+
+All application data is stored in PostgreSQL and object storage. If you have selected Mounted Disk mode, local services will be started for you, otherwise, the appropriate services will be configured to access a external PostgreSQL, S3-compatible blob storage, and, in active-active, Redis.
```
Comment (aaron-lane) [inline content/source/docs/enterprise/system-overview/architecture.html.md:49]: @amy-hashi What do you think about this suggestion to help clarify this section? I removed the mention of Redis because it is already discussed in the `Coordination Layer` section and the sentence seemed to imply that Redis was also part of the storage layer.

```suggestion
All application data is stored in a PostgreSQL database and an S3-compatible object storage service. If you have selected Mounted Disk mode, local services will be started for you; otherwise, the application must be configured to access an external database and object storage service.
```

**Review by aaron-lane — APPROVED** [review]
Thanks for incorporating my suggestions, @ml4.

LGTM 📜

**Comment by ml4** [comment]
Hello all, we're getting a little compression on internals from banks - would it be OK to merge it as it and update with ephemeral container list and update container names later please?

**Comment by laurapacilio** [comment]
Hi all - wanted to check in on this one, too! Is this something we'd want to merge in still/would we still want to make a few changes?

**Comment by aaron-lane** [comment]
@amy-hashi @sudomateo ☝🏼

**Comment by ml4** [comment]
There is still an increased level of interest primarily from global banks in knowing the internals as per the PR so that they can maintain decent comms with their security and compliance teams.

**Comment by laurapacilio** [comment]
Nice okay - I'm going to look through then to just do a quick review for flow, grammar, etc. @amy-hashi and @sudomateo would you mind also giving it a final thumbs up as well? Then we can get it merged in!

File: content/source/docs/enterprise/system-overview/architecture.html.md [hunk 10]
```diff
@@ -5,24 +5,201 @@ page_title: "Architecture Summary - System Overview - Terraform Enterprise"
...
+The application layer is divided conceptually into two areas, each with specific responsibilities:
+
+- Web Services - Answer requests forwarded by NginX and enqueue jobs for later
+- Background Services - Asynchronously process jobs enqueued by web services
+
+Most requests in Terraform Enterprise follow a similar lifecycle as shown in the diagram below.
+
+![tfe-int-app-layer](assets/tfe-int-app-layer.png)
+
+A web service receives a request and enqueues a background job to perform the task via the Coordination Layer (see below).  The web service then responds immediately to the original request instead of waiting for the job to be finished.
```
Comment (laurapacilio) [inline content/source/docs/enterprise/system-overview/architecture.html.md:25]:
```suggestion
A web service receives a request and enqueues a background job to perform the task via the Coordination Layer (see below). The web service then responds immediately to the original request instead of waiting for the job to be finished.
```

Comment (laurapacilio) [inline content/source/docs/enterprise/system-overview/architecture.html.md:25]: I might make these into numbered steps, since they're a progression and move them above the image.

File: content/source/docs/enterprise/system-overview/architecture.html.md [hunk 11]
```diff
@@ -5,24 +5,201 @@ page_title: "Architecture Summary - System Overview - Terraform Enterprise"
...
+
+## Storage Layer
+
+All application data is stored in PostgreSQL and object storage. If you have selected Mounted Disk mode, local services will be started for you, otherwise, the appropriate services will be configured to access a external PostgreSQL, S3-compatible blob storage, and, in active-active, Redis.
+
+~> **Note:** The configuration information provided and/or generated during installation (e.g. database credentials, hostname, etc.) is stored outside of the application.
+
+## Key Component Interactions
+
+This section details how the key containers work together to accomplish common tasks.
```
Comment (laurapacilio) [inline content/source/docs/enterprise/system-overview/architecture.html.md:55]: Are these the Docker containers? If so maybe we mention that?

File: content/source/docs/enterprise/system-overview/architecture.html.md [hunk 12]
```diff
@@ -5,24 +5,201 @@ page_title: "Architecture Summary - System Overview - Terraform Enterprise"
...
+
+### Initial Startup - External Services Mode
```
Comment (laurapacilio) [inline content/source/docs/enterprise/system-overview/architecture.html.md:57]: In all of these sections, I'd move the steps to be above the diagram. I think that will be more cohesive for users (to set them up for what they're observing in the diagram, and then we also won't split sentences that say "the following steps" from the following steps.

File: content/source/docs/enterprise/system-overview/architecture.html.md [hunk 13]
```diff
@@ -5,24 +5,201 @@ page_title: "Architecture Summary - System Overview - Terraform Enterprise"
```
Comment (laurapacilio) [inline content/source/docs/enterprise/system-overview/architecture.html.md:5]: Could you add a metadata description to the top of this page? 

description: |- 
  Learn about the functional layers of the architecture, how they interact, and details about the corresponding Docker containers.

File: content/source/docs/enterprise/system-overview/architecture.html.md [hunk 14]
```diff
@@ -5,24 +5,201 @@ page_title: "Architecture Summary - System Overview - Terraform Enterprise"
...
+2. Once `ptfe_nginx` has started, `rabbitmq` starts.
+3. Once `rabbitmq` has started, `postgres` starts.
+
+### Wait for Dependencies
+
+The application waits for `postgres`, `redis`, and `rabbitmq` to start as above. Once these key services respond successfully to requests, the startup process completes with the following steps.
+
+![tfe-int-wait-for-dependencies](assets/tfe-int-wait-for-dependencies.png)
+
+1. The remaining services that do not rely on PostgreSQL are launched in parallel. Included in this batch are or two short lived containers, `ptfe_migrations` and `ptfe_registry_migrations`, which exist only to apply any necessary updates to the database schema.<br />
```
Comment (laurapacilio) [inline content/source/docs/enterprise/system-overview/architecture.html.md:78]:
```suggestion
1. Terraform launches the remaining services that do not rely on PostgreSQL in parallel. These include two short-lived containers, `ptfe_migrations` and `ptfe_registry_migrations`, that exist only to apply any necessary updates to the database schema.<br />
```

File: content/source/docs/enterprise/system-overview/architecture.html.md [hunk 15]
```diff
@@ -5,24 +5,201 @@ page_title: "Architecture Summary - System Overview - Terraform Enterprise"
...
+2. The main **web** and **background** services start up:
+<br />2a. Once `ptfe_migrations` has finished, `ptfe_atlas` and `ptfe_sidekiq` are started in parallel.
+<br />2b. Once `ptfe_registry_migrations` has finished, `ptfe_registry_api` and `ptfe_registry_worker` are started in parallel.
+
+Finally, startup is complete and the application is ready to go at your configured hostname.
+
+### Module Ingress
+
+The process of importing Terraform modules into the private module registry starts with either a click in the UI or a `POST` to the API. The request is handled the same way in both cases:
```
Comment (laurapacilio) [inline content/source/docs/enterprise/system-overview/architecture.html.md:87]: Question (newbie here!) this "the private module registry," indicating that there's a universal private module registry. Is that the case, or is this "your private module registry."

File: content/source/docs/enterprise/system-overview/architecture.html.md [hunk 16]
```diff
@@ -5,24 +5,201 @@ page_title: "Architecture Summary - System Overview - Terraform Enterprise"
...
+
+![tfe-int-module-ingress](assets/tfe-int-module-ingress.png)
+
+1. `ptfe_atlas` receives request from `ptfe_nginx`, pushes a job to Redis for `ptfe_sidekiq`, and then responds to the original request immediately.
+1. `ptfe_sidekiq` asynchronously pulls the job and passes the details to `ptfe_slug_ingress`.
+1. `ptfe_slug_ingress` downloads your Git repository and `POST`s it to `ptfe_archivist`.
+1. `ptfe_archivist` persists the repository in blob storage and `POST`s the Object ID to `ptfe_atlas`.
+
+
+At this point, each tagged version of your repository has been persisted in the application. The process to make the modules available in your private module registry begins next:
```
Comment (laurapacilio) [inline content/source/docs/enterprise/system-overview/architecture.html.md:97]:
```suggestion
At this point, each tagged version of your repository has been persisted in the application. The process to make the modules available in your private module registry begins:
```

File: content/source/docs/enterprise/system-overview/architecture.html.md [hunk 17]
```diff
@@ -5,24 +5,201 @@ page_title: "Architecture Summary - System Overview - Terraform Enterprise"
...
 # Terraform Enterprise Architecture
 
-This page describes aspects of the architecture of Terraform Enterprise.
+Terraform Enterprise is powered by the same set of services behind HashiCorp’s public SaaS application Terraform Cloud. The sections below present the functional layers of the architecture, how they interact and details pertaining to the corresponding Docker containers involved. References to Docker containers in this document are those you will see when running a `docker ps` on a Terraform Enterprise application server while the application is up.
 
--> **Note:** This page was written for a prior implementation of Terraform Enterprise, and may be outdated. We plan to review our architecture documentation in the near future.
+~> **Advanced:** The content in this document is not required to operate Terraform Enterprise but is provided to help advanced users understand the product in more depth.
 
-## Services
+~> **Note:** Several services described below include the legacy string `ptfe` as part of their names. See [here](https://www.terraform.io/docs/enterprise/index.html#note-about-product-names) for more details.
```
Comment (sudomateo) [inline content/source/docs/enterprise/system-overview/architecture.html.md:12]: This is not technically a "legacy" string if there hasn't been an deprecation notice about replacing it. As much as I too want it gone.

File: content/source/docs/enterprise/system-overview/architecture.html.md [hunk 18]
```diff
@@ -5,24 +5,201 @@ page_title: "Architecture Summary - System Overview - Terraform Enterprise"
...
 
-These are the services used to run Terraform Enterprise. Each service contains a description of what actions it performs, a policy for restarts, impact of failing or degraded performance, and the service's dependencies.
+## Application Layer
+
+The application layer is divided conceptually into two areas, each with specific responsibilities:
+
+- Web Services - Answer requests forwarded by NginX and enqueue jobs for later
```
Comment (sudomateo) [inline content/source/docs/enterprise/system-overview/architecture.html.md:18]:
```suggestion
- Web Services - Answer requests forwarded by Nginx and enqueue jobs for later
```

File: content/source/docs/enterprise/system-overview/architecture.html.md [hunk 19]
```diff
@@ -5,24 +5,201 @@ page_title: "Architecture Summary - System Overview - Terraform Enterprise"
...
+
+1. `ptfe_atlas` sends the Object ID from `ptfe_archivist` to the `ptfe_registry_api`.
+1. `ptfe_registry_api` pulls the module from `ptfe_archivist`, saves the metadata in PostgreSQL, and passes the details to the `ptfe_registry_worker`.
+1. `ptfe_registry_worker` extracts the individual modules from your repository, updates the Private Registry tables in PostgreSQL, and `POST`s the results to `ptfe_atlas`.
+
+Finally, `ptfe_atlas` receives the callback and sets the module’s state to *ingressed* in PostgreSQL. The update is now live in the UI.
+
+### Webhook Receipt
+
+Communications between Terraform Enterprise and a connected VCS server happen via HTTP webhooks.  The latest state of the connected Git repository is downloaded using the process below, and then followup Terraform commands are applied by the state machine.
```
Comment (laurapacilio) [inline content/source/docs/enterprise/system-overview/architecture.html.md:109]: Just changing to active voice because it's easier to read - but please reject if this changes the meaning too much!

```suggestion
Communications between Terraform Enterprise and a connected VCS server happen via HTTP webhooks. Terraform downloads the latest state of the connected Git repository using the process below, and then the state machine applies the followup Terraform commands.
```

File: content/source/docs/enterprise/system-overview/architecture.html.md (hunk 9 above)
Comment (sudomateo) [reply content/source/docs/enterprise/system-overview/architecture.html.md:49]: +1 to this suggestion.

File: content/source/docs/enterprise/system-overview/architecture.html.md [hunk 20]
```diff
@@ -5,24 +5,201 @@ page_title: "Architecture Summary - System Overview - Terraform Enterprise"
...
+
+![tfe-int-coordination-layer](assets/tfe-int-coordination-layer.png)
+
+~> **Note:** While these components are critical to the overall operation of Terraform Enterprise, operators rarely (if ever) need to look at them directly.
+
+## Storage Layer
+
+All application data is stored in PostgreSQL and object storage. If you have selected Mounted Disk mode, local services will be started for you, otherwise, the appropriate services will be configured to access a external PostgreSQL, S3-compatible blob storage, and, in active-active, Redis.
+
+~> **Note:** The configuration information provided and/or generated during installation (e.g. database credentials, hostname, etc.) is stored outside of the application.
```
Comment (sudomateo) [inline content/source/docs/enterprise/system-overview/architecture.html.md:51]: Not sure what "outside" really means here.

File: content/source/docs/enterprise/system-overview/architecture.html.md (hunk 9 above)
Comment (aaron-lane) [reply content/source/docs/enterprise/system-overview/architecture.html.md:49]: @ml4 What do you think?

File: content/source/docs/enterprise/system-overview/architecture.html.md [hunk 21]
```diff
@@ -5,24 +5,201 @@ page_title: "Architecture Summary - System Overview - Terraform Enterprise"
...
+The application waits for `postgres`, `redis`, and `rabbitmq` to start as above. Once these key services respond successfully to requests, the startup process completes with the following steps.
+
+![tfe-int-wait-for-dependencies](assets/tfe-int-wait-for-dependencies.png)
+
+1. The remaining services that do not rely on PostgreSQL are launched in parallel. Included in this batch are or two short lived containers, `ptfe_migrations` and `ptfe_registry_migrations`, which exist only to apply any necessary updates to the database schema.<br />
+2. The main **web** and **background** services start up:
+<br />2a. Once `ptfe_migrations` has finished, `ptfe_atlas` and `ptfe_sidekiq` are started in parallel.
+<br />2b. Once `ptfe_registry_migrations` has finished, `ptfe_registry_api` and `ptfe_registry_worker` are started in parallel.
+
+Finally, startup is complete and the application is ready to go at your configured hostname.
```
Comment (sudomateo) [inline content/source/docs/enterprise/system-overview/architecture.html.md:83]: I believe we should remove this section entirely as this is subject to change and actually changing as we speak.

File: content/source/docs/enterprise/system-overview/architecture.html.md [hunk 22]
```diff
@@ -5,24 +5,201 @@ page_title: "Architecture Summary - System Overview - Terraform Enterprise"
...
+
+
+1. `ptfe_atlas` receives the callback request from `ptfe_nginx` and streams the response to `ptfe_archivist`.
+2. `ptfe_archivist` saves the Sentinel results to blob storage and returns an Object ID to `ptfe_atlas`.
+
+Finally, `ptfe_atlas` updates the metadata in PostgreSQL after it receives confirmation that the details of the run have been persisted.
+
+## Long-Running Container List
+
+The following table lists every container Terraform Enterprise runs via the Replicated Native Scheduler.  The list is presented by layer as per the above document sections and includes those long-running processes only, not ephemeral containers.
```
Comment (laurapacilio) [inline content/source/docs/enterprise/system-overview/architecture.html.md:157]:
```suggestion
The following table lists every container Terraform Enterprise runs via the Replicated Native Scheduler, following the layers outlined above. It does not include ephemeral containers, only those from long-running processes.
```

**Review by sudomateo — CHANGES_REQUESTED** [review]
I am not in favor of posting the container names. Instead, I would personally use the service names and how they interact with each other. For example, instead of saying that `ptfe_atlas` requests a request from `ptfe_nginx`, I would say the Atlas service receives a request from Nginx.

The container names may change over time, but the service names won't. Plus, in mounted disk mode many containers have `_disk` appended to the container name, making this document partially correct.

I also am not in favor of adding the long running container list to this page. While I think we should have more transparency around the containers that execute as part of Terraform Enterprise, I do not believe this page is the correct location for that content for reasons I mentioned above and because there are some ephemeral containers that run that do not have a name and are not listed here.

File: content/source/docs/enterprise/system-overview/architecture.html.md [hunk 23]
```diff
@@ -5,24 +5,201 @@ page_title: "Architecture Summary - System Overview - Terraform Enterprise"
...
+
+| Layer | Container Name | Process Name | Container Function |
+| - | - | - | - |
+| Web Services | `ptfe_atlas` | Atlas | Ruby on Rails application that powers the UI and API. |
+|  | `ptfe_registry_api` | Private Registry API | Go service that powers the private module registry. |
+|  | `ptfe_nginx` | Nginx | NginX is used as a router in Terraform Enterprise. |
+| Background Services | `ptfe_sidekiq` | Sidekiq | Ruby workers that run asynchronous jobs enqueued by `ptfe_atlas`. |
+|  | `ptfe_state_parser` | Terraform State Parser | Go service that parses Terraform State Files. |
+|  | `ptfe_build_manager` | Terraform Build Manager | Go service that coordinates Terraform runs, plan and applies. |
```
Comment (laurapacilio) [inline content/source/docs/enterprise/system-overview/architecture.html.md:166]:
```suggestion
|  | `ptfe_build_manager` | Terraform Build Manager | Go service that coordinates Terraform runs, plans, and applies. |
```

File: content/source/docs/enterprise/system-overview/architecture.html.md [hunk 24]
```diff
@@ -5,24 +5,201 @@ page_title: "Architecture Summary - System Overview - Terraform Enterprise"
...
+|  | `ptfe_registry_worker` | Private Module Registry Worker | Go service that parses Git repositories for Private Registry. |
```
Comment (laurapacilio) [inline content/source/docs/enterprise/system-overview/architecture.html.md:167]: I think Terraform Registry is a proper noun, but this isn't. Please correct if if I'm wrong!

```suggestion
|  | `ptfe_registry_worker` | Private Module Registry Worker | Go service that parses Git repositories for private registry. |
```

**Review by laurapacilio — CHANGES_REQUESTED** [review]

File: content/source/docs/enterprise/system-overview/architecture.html.md (hunk 17 above)
Comment (ml4) [reply content/source/docs/enterprise/system-overview/architecture.html.md:14]: I'll pull `legacy`

File: content/source/docs/enterprise/system-overview/architecture.html.md (hunk 20 above)
Comment (ml4) [reply content/source/docs/enterprise/system-overview/architecture.html.md:53]: I was separating Replicated from TFE here but I think your call is perspicacious, so I'll pull the comment.

File: content/source/docs/enterprise/system-overview/architecture.html.md (hunk 21 above)
Comment (ml4) [reply content/source/docs/enterprise/system-overview/architecture.html.md:85]: What is it changing to and when will the release happen? I'll comment `Wait for Dependencies`.

File: content/source/docs/enterprise/system-overview/architecture.html.md (hunk 15 above)
Comment (ml4) [reply content/source/docs/enterprise/system-overview/architecture.html.md:89]: PMR is organization-centric.

**Comment by ml4** [comment]
> I am not in favor of posting the container names. Instead, I would personally use the service names and how they interact with each other. For example, instead of saying that `ptfe_atlas` requests a request from `ptfe_nginx`, I would say the Atlas service receives a request from Nginx.
> 
> The container names may change over time, but the service names won't. Plus, in mounted disk mode many containers have `_disk` appended to the container name, making this document partially correct.
> 
> I also am not in favor of adding the long running container list to this page. While I think we should have more transparency around the containers that execute as part of Terraform Enterprise, I do not believe this page is the correct location for that content for reasons I mentioned above and because there are some ephemeral containers that run that do not have a name and are not listed here.
> 
> TL;DR Since this is an architecture overview page, I believe the content should stick with high-level service names instead of container names.

We are getting *explicit* requests from the field for a full list of container names as customers see them in a `docker ps` output and corresponding functions.  I have had another one from another global bank in the last day, and I'm sensing more than a degree of dissatisfaction about it, as technical teams feel they are somewhat less able to explain specific functionality to their compliance/security/audit teams about how the platform works due to the perceived paucity of documentation of this level.  

If we are to level up to our customer's expectations we need to absolutely list interactions as `ptfe_atlas` requests a request from `ptfe_nginx` _instead_ of Atlas service receives a request from Nginx, augmenting the documentation as necessary to ensure customers know that `ptfe_nginx` is part of, or the whole of, the Nginx service as necessary.  We then also need to list all of the containers that form part of the platform, and update them accordingly during releases in the same way we do other areas of the documentation.

I originally began this work to address customer concerns and used @glasner 's original document as a starting point as it had been heavily commented by Engineering in the regard so presumed it had pretty much already been agreed on.  With the deepest possible respect, I would propose we address the content so that it is not deemed partially correct but instead wholly correct, migrating areas of the content to other pages as needed and make it live as soon as possible.

**Comment by ihuntenator** [comment]
As a customer I really prefer this page here: https://github.com/hashicorp/terraform-website/blob/updateTFEarchitecture/content/source/docs/enterprise/system-overview/architecture.html.md
compared to what is currently here: https://www.terraform.io/docs/enterprise/system-overview/architecture.html
